Third-party-libraries.md

```diff
@@ -0,0 +1,255 @@
+3. Check for breaking changes in the outdated libraries from the previous step. You should:
+
+   * Check the library's changelog for breaking changes.
+   * Test that after you install the upgraded version (see below), everything works correctly.
+
+4. Manually update the versions in `package.json` for all packages you decide to upgrade.
+5. Run `../oppia_tools/yarn-<yarn version>/bin/yarn install`
+
+#### Upgrade frontend libraries not installed from npm
+
+Other frontend libraries are specified in `manifest.json`. You can upgrade these libraries as follows:
+
+1. Check for breaking changes in the libraries you want to upgrade. You should:
+
+   * Check the library's changelog for breaking changes.
+   * Test that after you install the upgraded version (see below), everything works correctly.
+
+2. Change the version in `manifest.json` to the new version you want to install.
+
+Note that we don't have an automatic way to detect outdated libraries in `manifest.json`, so to upgrade all such libraries, you have to check each version manually.
+
+## Add dependencies
+
+Note that all dependencies must be compatible with our Apache 2 license. All additions must also be approved by @vojtechjelinek.
+
+### Add a backend library
+
+Note that if a library is needed both for development and in production, then it should be added according to both of the following sets of instructions.
+
+#### Add a backend, production library from pip
+
+Here's how to add a backend, production library that can be installed from pip:
+
+1. Add the library to `requirements.in` in the form `<package-name>==<version>`.
+2. Generate `requirements.txt` from `requirements.in` by running `scripts/regenerate_requirements.py`.
+
+#### Add a backend, development library from pip
+
+To add a backend, development library that can be installed from pip, first ask whether the library is needed before running or importing any script. If the answer is yes, add it to the `PREREQUISITES` constant in `scripts/install_third_party_libs.py`. Otherwise, add it to the `local_pip_dependencies` variable in `scripts/install_third_party_libs.py`.
+
+#### Add a library not installable from pip
+
+If a library cannot be installed from pip, you'll have to add custom code to `scripts/install_third_party_libs.py` to handle installation.
+
+### Add a frontend library
+
+#### Add a frontend library from npm
+
+If the library is available from npm, you can install it like this:
+
+1. Add it to `package.json`. If the dependency is needed in production, add it under `"dependencies"`. Otherwise, add it under `"devDependencies"`. If a dependency is needed in both production and for development, only add it under `"dependencies"`.
+2. Run `../oppia_tools/yarn-<yarn version>/bin/yarn install` from the repository root. Note that `<yarn version>` specifies the version of yarn.
+
+#### Add a frontend library from outside npm
+
+If your library is not available from npm, you can add it to `manifest.json`.
```

Comment:
This conflicts with "manifest.json is deprecated" above.

Reply:
I think the word "deprecated" was confusing, so I removed it. By "deprecated" I meant that we (Oppia) wanted to move away from this method of installing packages. `manifest.json` is our own creation, so we're the ones "deprecating" it. However, I'm leaving open the possibility that we might need a frontend dependency that's not from npm, in which case we'll have to keep `manifest.json` around. To reflect that possibility, I've made the "don't use manifest.json" guidance a little weaker